x/mongo/driver/crypt.go

```diff
@@ -382,3 +389,69 @@ func (c *crypt) decryptKey(kmsCtx *mongocrypt.KmsContext) error {
 		}
 	}
 }
+
+// needsKmsProvider returns true if provider was initially set to an empty document.
+// An empty document signals the driver to fetch credentials.
+func needsKmsProvider(kmsProviders *bsoncore.Document, provider string) bool {
+	gcpVal, err := kmsProviders.LookupErr(provider)
+	if err != nil {
+		// KMS provider is not configured.
+		return false
+	}
+	gcpDoc, ok := gcpVal.DocumentOK()
+	if !ok || len(gcpDoc) != 5 {
+		// KMS provider is not an empty document.
+		return false
+	}
+	return true
+}
+
+func (c *crypt) provideKmsProviders(cryptCtx *mongocrypt.Context) error {
+	kmsProviders := c.mongoCrypt.GetKmsProviders()
+	builder := bsoncore.NewDocumentBuilder()
+
+	if needsKmsProvider(kmsProviders, "gcp") {
+		// "gcp" KMS provider is an empty document.
+		// Attempt to fetch from GCP Instance Metadata server.
+		{
+			metadataHost := "metadata.google.internal"
+			if os.Getenv("GCE_METADATA_HOST") != "" {
+				metadataHost = os.Getenv("GCE_METADATA_HOST")
+			}
+			url := fmt.Sprintf("http://%s/computeMetadata/v1/instance/service-accounts/default/token", metadataHost)
+			req, err := http.NewRequest("GET", url, nil)
```

Comment:
To allow the calling code to cancel this blocking function, add a context argument to the `provideKmsProviders` function and create this HTTP request with a context.

E.g.
```go
func (c *crypt) provideKmsProviders(ctx context.Context, cryptCtx *mongocrypt.Context) error {
	// ...
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	// ...
```